[prototype] server,rpc: host a dRPC server and use it for node-node BatchRequest

tbg opened the pull request:
Informs https://github.com/cockroachdb/cockroach/issues/137252.

```
$ benchdiff --old lastmerge ./pkg/sql/tests -b -r 'Sysbench/SQL/3node/oltp_read_write' -d 1000x -c 10

name                                   old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_read_write-24    14.0ms ± 2%    14.1ms ± 1%     ~     (p=0.063 n=10+10)

name                                   old alloc/op   new alloc/op   delta
Sysbench/SQL/3node/oltp_read_write-24    2.55MB ± 1%    2.26MB ± 2%  -11.39%  (p=0.000 n=9+9)

name                                   old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_read_write-24     17.2k ± 1%     11.7k ± 0%  -32.30%  (p=0.000 n=10+8)
```

This commit sets up the plumbing to host a drpc server alongside our
usual gRPC server.

Readers may find [1] a good resource to understand the broad strokes
of what the essence of this approach looks like before diving into
this PR.

This PR has a few parts:

- it adds a copy of drpc generated code for a service that supports
  unary BatchRequest.
- in `rpc.NewServerEx`, we now also set up a drpc server and return
  it up the stack. (This is somewhat awkwardly done by hiding the
  drpc server inside a `grpcServer` struct, but that should be easy
  to fix with more plumbing or a rename).
- that server is registered for the (drpc) BatchRequest service.
- in our listener setup, we configure `cmux` to match on the `DRPC!!!1`
  header and serve the resulting listener on the drpc server. The drpc
  example uses `drpcmigrate.ListenMux` instead of `cmux`; we keep `cmux`
  but must make sure the header is read off the connection before
  delegating (which the `drpxmigrate` mux would have done for us).
- if using TLS, wrap the resulting listener.
- then, serve it using the drpc server.
- I added a unit test that sets up a TestCluster and demonstrates that we
  can manually make a connection, set up a drpc client, and send a
  simple BatchRequest. In particular, this cluster "just works", meaning
  that the mux works - gRPC continues to be functional and is blissfully
  unaware that you can also send drpc requests to our listener.

What's missing is the ability for nodes to intrinsically use drpc. This
will require some more work in the `rpc.Context` where we associate with
each `*Connection` a `*drpcpool.Pool` of connections. We then need to
change `kvcoord/transport.go` to use these drpc conns. This will largely
parallel what https://github.com/cockroachdb/cockroach/pull/136648 already does,
so I can mostly copy the approach there. Also, once that PR merges, I should be
able to also adopt a mode where we reuse batch streams for drpc, too, which we
know to have benefits.

cc @nvanbenschoten @cthumuluru-crdb 

[1]: https://github.com/storj/drpc/tree/main/examples/grpc_and_drpc

tbg committed:
server,rpc: host a dRPC server

This commit sets up the plumbing to host a drpc server alongside our
usual gRPC server. The server is hosted only if the
`rpc.experimental_drpc.enabled` cluster setting _can be enabled_,
which is the case either in CRDB test builds or when the env var
COCKROACH_EXPERIMENTAL_DRPC_ENABLED is set.

Readers may find [1] a good resource to understand the broad strokes
of what the essence of this approach looks like before diving into
this PR.

This PR has a few parts:

- it adds a copy of drpc generated code for a service that supports
  unary BatchRequest.
- in `rpc.NewServerEx`, we now also set up a drpc server and return
  it up the stack. (This is somewhat awkwardly done by hiding the
  drpc server inside a `grpcServer` struct, but that should be easy
  to fix with more plumbing or a rename).
- that server is registered for the (drpc) BatchRequest service.
- in our listener setup, we configure `cmux` to match on the `DRPC!!!1`
  header and serve the resulting listener on the drpc server. The drpc
  example uses `drpcmigrate.ListenMux` instead of `cmux`; we keep `cmux`
  but must make sure the header is read off the connection before
  delegating (which the `drpxmigrate` mux would have done for us).
- if using TLS, wrap the resulting listener.
- then, serve it using the drpc server.
- I added a unit test that sets up a TestCluster and demonstrates that we
  can manually make a connection, set up a drpc client, and send a
  simple BatchRequest. In particular, this cluster "just works", meaning
  that the mux works - gRPC continues to be functional and is blissfully
  unaware that you can also send drpc requests to our listener.

What's missing is the ability for nodes to intrinsically use drpc. This
will require some more work in the `rpc.Context` where we associate with
each `*Connection` a `*drpcpool.Pool` of connections. We then need to
change `kvcoord/transport.go` to use these drpc conns. I expect this
all to be rather messy, but given what this PR achieves already there
is little doubt that it can be done.

[1]: https://github.com/storj/drpc/tree/main/examples/grpc_and_drpc

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136794)
<!-- Reviewable:end -->

tbg committed:
rpc,kv: use drpc for BatchRequest

I cherry-picked the stream reuse PR #136648 so that both are
more likely to evolve in unison, and because I anticipated
"piggybacking" on top of it. This did not materialize, but
the important integration point is now side by side and should
lend itself well to a future change that adds stream reuse for
drpc as well.

This commit roughly encompasses the following:

1.  we change `*rpc.Connection` to also maintain a `drpcpool.Conn`.

It took me a while to grok how the `drpcpool.Pool` is architected.
Essentially, its `Get()` method gives you a handle to a `drpcpool.Conn`
that reflects a specific use case of the pool. When a client is created
on it, it gets assigned an actual `drpc.Conn` from the pool (dialing if
necessary), and when the client is closed, this conn is returned to the
pool.

So in a sense, `drpcpool.Conn` is an actual pool for some keyed use
case; `drpcpool.Pool` is simply the bucket in which the actual
connections get pooled, but they're never pulled from or returned to it
directly. We don't currently use the key since we create a pool per
remote node, and if we're not sharing TCP conns they all look the same
to us anyway (i.e. there's no point in DefaultClass vs SystemClass).

If we squint, a `drpcoool.Conn` parallels `*grpc.ClientConn` in the sense
that you can "just" make multiple clients on top of it.  Of course
internally a `*grpc.ClientConn` multiplexes multiple clients over one
HTTP2 connection, whereas a `drpcpool.Conn` represents multiple
independent TCP connections.

The lifecycle of these pools is currently completely broken, and I was
honestly surprised TestTestServerDRPC passes! Future commits will need
to clean this up to the point where we can at least run a representative
benchmark.

tbg committed:
wrap conns in pool in a closeConnOnWriteFailureConn

Saw the following:

sysbench prepare segfaults
see that index backfill jobs don't all finish and take a look time
use the job observability (good stuff!) to get a sense of what stack traces the job has
see one that's in DistSender retry loop
logspy into the node, see following error:

```
 I241206 16:07:08.257987 120483 kv/kvclient/kvcoord/dist_sender.go:2246 ⋮ [T1,Vsystem,n2,f‹81eb2d54›,job=0,distsql.gateway=3,distsql.appname=‹$ internal-plan-index-backfill›,indexBackfillerProcessor=112] 14561 evicting range desc ‹et:desc:r163:/Table/112/{2-3} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=16, sticky=1733502749.098515116,0], lease:<empty>› after ‹failed to send RPC›: sending to all replicas failed; last error: ba: ‹AddSSTable [/Table/112/2/1/712843/0,/Table/112/2/1958386/1707563/0/NULL)› RPC error: ‹write tcp 10.142.1.75:59158->10.142.1.95:26257: write: connection reset by peer; use of closed network connection›
```

ssh into the machine, no outgoing connection from port 59158 so this conn is dead.
read connpool code, see that it isn't concerned with the conn's health at all. someone else would have to close the conn.
looked at goroutines for stream.manager; hard to tell which is for which conn but they're all just on a long-poll read and
i'm not even sure the one I care about is even still there (maybe it terminated but somehow left the conn looking healthy
enough? though I doubt that)

this commit is really more of a hack. We'll clearly need a better story here.

Also no idea why the conn would break. The cluster was just started, maybe n2 was reaching out to n3 before it was up
but then how did the conn manager not notice.

needs unit testing for sure.

tbg committed:
set ReaderOptions.MaximumBufferSize

Couldn't `sysbench prepare`, it'd always get stuck on the index addition
jobs. Saw message like these:

 E241206 22:39:10.777982 1418 rpc/context.go:197 ⋮ [-] 174  TBG server error ‹manager closed›: ‹protocol error›: ‹data overflow (len:4259840)›

That's just over 4mb. Realized there is a default value for
`ReaderOptions.MaximumBufferSize` of 4mb, so these AddSST calls would
never succeed.

This commit sets that limit to infinity. Will have to check that this
doesn't cause large memory consumption if the buffer grows once. Will
it reduce in size again? I think I saw code like this but need to check.

There's also a StreamOptions.SplitSize. It's probably better to keep the
max read buffer size limited but set a SplitSize of a couple of MBs
worth. Trying that out is for another commit though, today I just want
something that works!

This commit also backs out some of the "close on rpc error" logic since
it was just adding a lot of spam and now that I understand what the
problem was was misguided. We weren't accidentally keeping broken
connections in the pool. We were constantly making and breaking
connections, I think.

tbg committed:
nodedialer,rpc,kvserver: add drpc stream pooling

This enables use of drpc stream pooling, gated by the same cluster
setting and version that is used for gRPC stream pooling.

Noting that `drpcpool.Conn` (which is what we use) implements a TCP
connection pool under the hood, we can in effect reuse the entire gRPC
batch streaming infrastructure without changes because both
`*grpc.ClientConn` and `drpcconn.Conn` can be used for multiple
concurrent streams (the internal difference being that in the latter
case, the streams won't be sharing TCP connections).

```
name                                   old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_read_write-24    13.4ms ± 2%    12.5ms ± 2%  -6.73%  (p=0.000 n=10+10)

name                                   old alloc/op   new alloc/op   delta
Sysbench/SQL/3node/oltp_read_write-24    2.28MB ± 0%    2.11MB ± 1%  -7.41%  (p=0.000 n=9+8)

name                                   old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_read_write-24     11.7k ± 0%     11.5k ± 1%  -1.73%  (p=0.000 n=9+9)
```

tbg committed:
kvserver: extract (*Node).batchStreamImpl

This is so we can reuse it for both gRPC and drpc.

tbg committed:
rpc: add comment in NewServerEx about drpc server interceptors

tbg committed:
cleanup: remove closeConnOnWriteFailure etc in peer

tbg committed:
rpc: add rpc.experimental_drpc.enabled cluster setting

tbg committed:
[dnm] rpc: COCKROACH_EXPERIMENTAL_DRPC_ENABLED=true

tbg committed:
fixup! rpc,kv: use drpc for BatchRequest

cthumuluru-crdb commented on pkg/rpc/nodedialer/nodedialer.go:
```diff
@@ -147,18 +148,39 @@ func (n *Dialer) DialInternalClient(
 	client := newBaseInternalClient(conn)
+
+	const useDRPC = true
+	if useDRPC {
```
Overwrite the client only when drpc is enabled.

cthumuluru-crdb replied:
nvm, I see this already fixed in followup commits.

cthumuluru-crdb reviewed (commented):
Thanks for the changes, @tbg. I reviewed most part of the code and left some comments. I would like to take another pass at it before committing these changes.
